mongodbreceiver: index-stats scrape errors on views and timeseries collections (Location40602)

### Component(s)

receiver/mongodb

### What happened?

**Description**

The index-stats scraper enumerates namespaces with `ListCollectionNames` (no filter) and runs `$indexStats` against each one. Views (and timeseries collections) reject `$indexStats`, so every scrape cycle logs one error per view:

```
error scraperhelper/obs_metrics.go:61 Error scraping metrics
{"otelcol.component.id": "mongodb/...", "scraper": "mongodb",
 "error": "failed to fetch index stats metrics: (Location40602) $indexStats is only valid as the first stage in a pipeline"}
```

The error is misleading — `$indexStats` *is* the first stage in the receiver's pipeline, but when the target is a view, MongoDB prepends the view's own definition pipeline, displacing it.

**Steps to Reproduce**

1. `db.createView("v", "someCollection", [{ $match: {} }])` in any monitored database
2. Run the collector with the `mongodb` receiver (default metrics; `mongodb.index.access.count` enabled)
3. One Location40602 scrape error per view per `collection_interval`

**Expected Result**

Views and timeseries collections skipped for index stats (they cannot have index stats); no scrape errors.

**Actual Result**

Persistent scrape errors; index stats for regular collections are still collected, so the errors are noise, but they fire every cycle and cannot be suppressed without disabling `mongodb.index.access.count` entirely.

**Proposed fix**

Filter the enumeration to real collections — `ListCollectionNames` is only used by the index-stats path:

```go
func (c *mongodbClient) ListCollectionNames(ctx context.Context, database string) ([]string, error) {
	lcOpts := options.ListCollections().SetAuthorizedCollections(true)
	return c.Database(database).ListCollectionNames(ctx, bson.D{{Key: "type", Value: "collection"}}, lcOpts)
}
```

We are running this patch in production against MongoDB Atlas 8.x replica sets: errors gone, `mongodb.index.access.count` unaffected for regular collections. Happy to open a PR with the change + changelog if maintainers agree with the approach.

### Collector version

v0.155.0

### Environment information

Custom OCB distribution (linux/amd64), MongoDB Atlas M30 replica sets (MongoDB 8.x), mongodb+srv over PrivateLink.

--- a/.chloggen/dylan_ignore-view-index-scrape-errors.yaml
+++ b/.chloggen/dylan_ignore-view-index-scrape-errors.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. receiver/filelog)
+component: receiver/mongodb
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Ignore the `$indexStats` scrape error that occurs when scraping views
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [49494]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/receiver/mongodbreceiver/scraper.go
+++ b/receiver/mongodbreceiver/scraper.go
@@ -610,7 +610,7 @@ func (s *mongodbScraper) collectIndexStats(ctx context.Context, now pcommon.Time
 	indexStats, err := s.client.IndexStats(ctx, databaseName, collectionName)
 	if err != nil {
 		var srvErr mongo.ServerError
-		// ignore error $indexStats is not supported on views
+		// ignore the error, $indexStats is not supported on views
 		if errors.As(err, &srvErr) && srvErr.HasErrorCode(40602) {
 			s.logger.Debug("skipping unsupported index stats for view",
 				zap.String("database", databaseName), zap.String("view", collectionName))

--- a/receiver/mongodbreceiver/scraper.go
+++ b/receiver/mongodbreceiver/scraper.go
@@ -15,6 +15,7 @@ import (
 	"github.com/google/uuid"
 	"github.com/hashicorp/go-version"
 	"go.mongodb.org/mongo-driver/v2/bson"
+	"go.mongodb.org/mongo-driver/v2/mongo"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/config/confignet"
 	"go.opentelemetry.io/collector/pdata/pcommon"
@@ -608,6 +609,13 @@ func (s *mongodbScraper) collectIndexStats(ctx context.Context, now pcommon.Time
 	}
 	indexStats, err := s.client.IndexStats(ctx, databaseName, collectionName)
 	if err != nil {
+		var srvErr mongo.ServerError
+		// ignore error $indexStats is not supported on views
+		if errors.As(err, &srvErr) && srvErr.HasErrorCode(40602) {
+			s.logger.Debug("skipping unsupported index stats for view",
+				zap.String("database", databaseName), zap.String("view", collectionName))
+			return
+		}
 		errs.AddPartial(1, fmt.Errorf("failed to fetch index stats metrics: %w", err))
 		return
 	}

--- a/receiver/mongodbreceiver/scraper_test.go
+++ b/receiver/mongodbreceiver/scraper_test.go
@@ -1675,3 +1675,14 @@ func TestDependentMetricsWhenDisabled(t *testing.T) {
 		})
 	}
 }
+
+func TestCollectIndexStatsSkipsViews(t *testing.T) {
+	fc := &fakeClient{}
+	viewErr := mongo.CommandError{Code: 40602, Message: "$indexStats is only valid as the first stage in a pipeline"}
+	fc.On("IndexStats", mock.Anything, "fakedatabase", "someview").Return([]bson.M{}, viewErr)
+	scraper := &mongodbScraper{logger: zap.NewNop(), client: fc}
+
+	errs := &scrapererror.ScrapeErrors{}
+	scraper.collectIndexStats(t.Context(), pcommon.NewTimestampFromTime(time.Now()), "fakedatabase", "someview", errs)
+	require.NoError(t, errs.Combine())
+}

